File: src/core/udp/services/peer-session-service.ts
import type { PeerSession } from "../types/udp-type.ts";

/** Sessions idle longer than this are dropped (join retries far exceed it). */
const SESSION_IDLE_TIMEOUT_MS = 60_000;
const SWEEP_INTERVAL_MS = 10_000;

/** Owns and tracks negotiated p2p sessions, keyed by remote endpoint. */
export class PeerSessionService {
  private readonly sessions = new Map<string, PeerSession>();
  private sweepTimer: number | null = null;

  get(remoteAddress: string): PeerSession | null {
    return this.sessions.get(remoteAddress) ?? null;
  }

  put(session: PeerSession): void {
    this.sessions.set(session.remoteAddress, session);
  }

  delete(remoteAddress: string): void {
    this.sessions.delete(remoteAddress);
  }

  /** Re-key a session when a peer's source endpoint changes (rare). */
  rename(oldKey: string, session: PeerSession): void {
    if (oldKey !== session.remoteAddress) {
      this.sessions.delete(oldKey);
    }
    this.sessions.set(session.remoteAddress, session);
  }

  start(): void {
    if (this.sweepTimer !== null) return;
    this.sweepTimer = setInterval(() => this.sweep(), SWEEP_INTERVAL_MS) as unknown as number;
  }

  stop(): void {
    if (this.sweepTimer !== null) {
      clearInterval(this.sweepTimer);
      this.sweepTimer = null;
    }
    this.sessions.clear();
  }

  private sweep(): void {
    const now = Date.now();
    for (const [key, session] of this.sessions) {
      if (now - session.lastSeenAt > SESSION_IDLE_TIMEOUT_MS) {
        this.sessions.delete(key);
      }
    }
  }
}

File: src/Shared/Udp/PeerSessionService.cs
using Mgo2Server.Shared.Types;

namespace Mgo2Server.Shared.Udp;

/// <summary>
/// Holds the negotiated peer-to-peer sessions, keyed by remote endpoint. The
/// sessions are reaped once they have gone quiet, which releases the peers that
/// never completed a handshake.
/// </summary>
/// <param name="idleTimeout">How long a session may stay quiet before it is dropped.</param>
public sealed class PeerSessionService(TimeSpan idleTimeout)
{
    private readonly Lock gate = new();
    private readonly Dictionary<string, PeerSession> sessions = [];
    private Timer? reaper;

    /// <summary>Starts the reaper that drops idle sessions.</summary>
    public void Start() =>
        reaper ??= new Timer(_ => Reap(), null, TimeSpan.FromSeconds(15), TimeSpan.FromSeconds(15));

    /// <summary>Stops the reaper.</summary>
    public void Stop()
    {
        reaper?.Dispose();
        reaper = null;
    }

    /// <summary>Returns the session of a remote endpoint, when it has one.</summary>
    /// <param name="remoteAddress">Endpoint formatted as "address:port".</param>
    public PeerSession? Get(string remoteAddress)
    {
        lock (gate)
        {
            return sessions.GetValueOrDefault(remoteAddress);
        }
    }

    /// <summary>Stores a session, replacing any earlier one for the endpoint.</summary>
    /// <param name="session">Session to store.</param>
    public void Put(PeerSession session)
    {
        lock (gate)
        {
            sessions[session.RemoteAddress] = session;
        }
    }

    /// <summary>Drops the session of a remote endpoint.</summary>
    /// <param name="remoteAddress">Endpoint formatted as "address:port".</param>
    public void Remove(string remoteAddress)
    {
        lock (gate)
        {
            sessions.Remove(remoteAddress);
        }
    }

    /// <summary>Returns a snapshot of every live session.</summary>
    public List<PeerSession> Snapshot()
    {
        lock (gate)
        {
            return [.. sessions.Values];
        }
    }

    private void Reap()
    {
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        lock (gate)
        {
            foreach (var (remoteAddress, session) in sessions.ToList())
            {
                if (now - session.LastSeenAt > idleTimeout.TotalMilliseconds)
                {
                    sessions.Remove(remoteAddress);
                }
            }
        }
    }
}
